File: plugins/distortion/saturator/saturator.js
// Saturator — soft-clip distortion with drive + dry/wet mix.
// AudioWorklet-style plugin written in ES5 prototype syntax (the host
// JS engine on mobile doesn't parse ES6 `class`). The contract is the
// same: registerProcessor('id', Ctor) where Ctor has a `process` method
// and an optional `parameterDescriptors` static array.

function SaturatorProcessor() {}
SaturatorProcessor.parameterDescriptors = [
  { name: 'drive', label: 'Drive', defaultValue: 0.5, minValue: 0, maxValue: 1 },
  { name: 'mix',   label: 'Mix',   defaultValue: 1.0, minValue: 0, maxValue: 1 }
];
SaturatorProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0];
  var output = outputs[0][0];
  if (!input || !output) return true;
  var drive = parameters.drive[0];
  var mix = parameters.mix[0];
  var k = 1 + drive * 8;
  for (var i = 0; i < input.length; i++) {
    var wet = Math.tanh(input[i] * k);
    output[i] = input[i] * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('saturator', SaturatorProcessor);

File: plugins/distortion/saturator/Saturator.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Saturator (native port) — soft-clip tube saturation via tanh. Drive
// pushes the input deeper into the curve; mix blends back the dry path
// so even extreme drive can sit in a mix.
public final class Saturator implements VocalMonitorNativePlugin {
    private float drive = 0.5f, mix = 1f;

    @Override public void init(int sr) {}
    @Override public String[] parameterNames() { return new String[] { "drive", "mix" }; }
    @Override public float parameterMin(String n) { return 0f; }
    @Override public float parameterMax(String n) { return 1f; }
    @Override public float parameterDefault(String n) { return "drive".equals(n) ? 0.5f : 1f; }
    @Override public String parameterLabel(String n) { return "drive".equals(n) ? "Drive" : "Mix"; }
    @Override public void setParameter(String n, float v) {
        if ("drive".equals(n)) drive = v;
        else if ("mix".equals(n)) mix = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final float k = 1f + drive * 8f;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float wet = (float) Math.tanh(input[i] * k);
            output[i] = input[i] * dry + wet * mixLocal;
        }
    }
}
